README.md
# Image Modernizer
Welcome to the Image Modernizer Repository! This repo is the home for Image Modernizer - a desktop software to modernize your images.

## Motivation
A few weeks ago, I checked out my website using AHREF's site audit tool and realized my PNG images were too big.
I googled how to fix this problem and resizing the images was not the most optimal solution - modernizing them to effective image formats was.

But how do you do that?

I googled that too and voila - I found a plethora of online converters. Just upload the image and be done with it.
But upload to where? Who's the owner of that site? What if I decide to operate an internal website and I don't want to upload the pictures to other servers?

There's absolutely no guarantee that website owners don't keep your images. They can say that they don't but you don't know for sure.
So, I decided to build my own open source converter. This is a desktop app that makes no requests to the internet and uses a very simple UI to convert your local images to WEBP - a modern image format that googlebot would love to see when it scans your website.

## How to use Image Modernizer?
Go to the [releases](https://github.com/orencohendev/image-modernizer/releases) page and download the software from the latest release.
Three platforms are currently supported: Windows, MacOS, and Linux.
Run Image Modernizer on the system you're using and you'll see a simple UI with a "Browse" button.
Select your PNG or JPEG image and let Image Modernizer do the rest. When done, it will ask you to select the destination for the new Webp copy of your image.

Enjoy!

## Next Steps
I don't have a lot of expertise making desktop apps but I do have vast knowledge in using Python as a backend engineer, and an open mind to learn more.
If you're interested in contributing to this project by adding more input and output formats, be my guest!

If you have a suggestion or encountered a bug, feel free to open an issue.

Thanks for checking out this repository, and I hope it helps you!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.1]

### Fixes

- CI: The workflow artifact for macos and linux had the same name.


## [1.1.0]

### Added

- Added NSIS installer for the Windows build.

## [1.0.0]

### Added

- Rebranded the software as Image Modernizer.
- Added a theme to make the UI more inviting.
- Added support for JPEG and JPG files.

### Fixed

- Removed the console window that was opening when launching the built binary by adding `--noconsole` to pyinstaller.
- Made sure the original file name carried over to the webp save file dialog.

## [0.1.1] - 2023-06-14

### Added

- support for windows and mac releases.
- initial release

### fixed

- changes to CI yaml to correctly build.


png_to_webp_converter.py
import os
import tkinter as tk
from tkinter import filedialog, messagebox as alert
from tkinter import ttk
from PIL import Image
from threading import Thread
import time

class Application(tk.Frame):
    def __init__(self, master=None):
        super().__init__(master)
        self.master = master
        self.master.title("Welcome to PNG Converter")
        self.pack()
        self.create_widgets()

    def create_widgets(self):
        self.title = tk.Label(self, text="Welcome to PNG Converter", font=("Arial", 20))
        self.title.pack(side="top")

        self.subtitle = tk.Label(self, text="Please click the Browse button to select your file", font=("Arial", 14))
        self.subtitle.pack(side="top")
        
        self.empty_line = tk.Label(self, text="")
        self.empty_line.pack(side="top")

        self.browse = tk.Button(self, text="Browse", font=("Arial", 20), height=2, width=10)
        self.browse["command"] = self.open_file_dialog
        self.browse.pack(side="top")

        self.progress = ttk.Progressbar(self, length=200, mode='indeterminate')

    def open_file_dialog(self):
        filepath = filedialog.askopenfilename(filetypes=[("PNG Files", "*.png"), ("JPEG Files", "*.jpg"), ("JPEG Files", "*.jpeg")])
        if filepath:
            self.progress.pack(side="bottom")
            self.progress.start()
            Thread(target=self.convert_image, args=(filepath,)).start()


    def get_filename_without_extension(self, path):
        base_name = os.path.basename(path)  # Get the filename from path
        file_name_without_ext = os.path.splitext(base_name)[0]  # Get the filename without extension
        return file_name_without_ext


    def convert_image(self, filepath):
        try:
            img = Image.open(filepath)

            time.sleep(1)  # Simulate processing time.
            self.progress.stop()
            self.progress.pack_forget()
            save_path = filedialog.asksaveasfilename(initialfile=self.get_filename_without_extension(filepath),defaultextension=".webp", filetypes=[("WebP Files", "*.webp")])
            if save_path:
                img.save(save_path, "WEBP")
                alert.showinfo(title="Conversion Complete!",message="Image converted and saved successfully.")
        except Exception as e:
            alert.showerror(title="Conversion Failed!",message=f"Error occurred while converting image: {str(e)}")

def run():
    root = tk.Tk()
    app = Application(master=root)
    app.mainloop()

if __name__ == '__main__':
    run()


image_modernizer.py
import os
import tkinter as tk
from tkinter import filedialog, messagebox as alert
from ttkbootstrap import Style, ttk
from PIL import Image
from threading import Thread
import time


class Application(tk.Frame):
    def __init__(self, master=None):
        super().__init__(master)
        self.master = master
        self.master.title("Welcome to Image Modernizer")
        self.grid()
        self.create_widgets()

    def create_widgets(self):
        self.title = tk.Label(
            self, text="Welcome to Image Modernizer", font=("Open Sans", 20)
        )
        self.title.grid(row=0, column=0, columnspan=2, padx=10, pady=10)

        self.subtitle = tk.Label(
            self,
            text="Please click the Browse button to select the\n file you want to modernize or click exit.",
            font=("Open Sans", 14),
        )
        self.subtitle.grid(row=1, column=0, columnspan=2, padx=10, pady=10)

        self.empty_line = tk.Label(self, text="")
        self.empty_line.grid(row=2, column=0, columnspan=2, padx=10, pady=10)

        self.browse = ttk.Button(self, text="Browse", style="info.TButton")
        self.browse["command"] = self.open_file_dialog
        self.browse.grid(row=3, column=0, padx=10, pady=10)

        self.exit = ttk.Button(self, text="Exit", style="danger.TButton")
        self.exit["command"] = self.exit_gracefully
        self.exit.grid(row=3, column=1, padx=10, pady=10)

        self.progress = ttk.Progressbar(self, length=200, mode="indeterminate")

    def exit_gracefully(self):
        self.master.destroy()

    def open_file_dialog(self):
        filepath = filedialog.askopenfilename(
            filetypes=[
                ("PNG Files", "*.png"),
                ("JPEG Files", "*.jpg"),
                ("JPEG Files", "*.jpeg"),
            ]
        )
        if filepath:
            self.progress.grid(row=4, column=0, columnspan=2, padx=10, pady=10)
            self.progress.start()
            Thread(target=self.convert_image, args=(filepath,)).start()

    def get_filename_without_extension(self, path):
        base_name = os.path.basename(path)  # Get the filename from path
        file_name_without_ext = os.path.splitext(base_name)[
            0
        ]  # Get the filename without extension
        return file_name_without_ext

    def convert_image(self, filepath):
        try:
            img = Image.open(filepath)

            time.sleep(1)  # Simulate processing time.
            self.progress.stop()
            self.progress.grid_forget()
            save_path = filedialog.asksaveasfilename(
                initialfile=self.get_filename_without_extension(filepath),
                defaultextension=".webp",
                filetypes=[("WebP Files", "*.webp")],
            )
            if save_path:
                img.save(save_path, "WEBP")
                alert.showinfo(
                    title="Conversion Complete!",
                    message="Image converted and saved successfully.",
                )
        except Exception as e:
            alert.showerror(
                title="Conversion Failed!",
                message=f"Error occurred while converting image: {str(e)}",
            )


def run():
    style = Style(theme="darkly")
    root = style.master
    app = Application(master=root)
    app.mainloop()


if __name__ == "__main__":
    run()
